src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/FileBasedPrograms/VirtualProjectXmlDiagnosticSourceProvider.cs

```diff
@@ -0,0 +1,91 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Composition;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
+using Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.FileBasedPrograms;
+
+[Export(typeof(IDiagnosticSourceProvider)), Shared]
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal class VirtualProjectXmlDiagnosticSourceProvider(VirtualProjectXmlProvider virtualProjectXmlProvider) : IDiagnosticSourceProvider
+{
+    public bool IsDocument => true;
+
+    public const string FileBasedPrograms = nameof(FileBasedPrograms);
+    public string Name => FileBasedPrograms;
+
+    public bool IsEnabled(ClientCapabilities clientCapabilities) => true;
+
+    public ValueTask<ImmutableArray<IDiagnosticSource>> CreateDiagnosticSourcesAsync(RequestContext context, CancellationToken cancellationToken)
+    {
+        ImmutableArray<IDiagnosticSource> sources = context.Document is null
+            ? []
+            : [new DiagnosticSource(context.Document, virtualProjectXmlProvider)];
+
+        return ValueTask.FromResult(sources);
+    }
+
+    private class DiagnosticSource(Document document, VirtualProjectXmlProvider virtualProjectXmlProvider) : IDiagnosticSource
+    {
+        public async Task<ImmutableArray<DiagnosticData>> GetDiagnosticsAsync(RequestContext context, CancellationToken cancellationToken)
+        {
+            if (string.IsNullOrEmpty(document.FilePath))
+                return [];
+
+            var simpleDiagnostics = await virtualProjectXmlProvider.GetCachedDiagnosticsAsync(document.FilePath, cancellationToken);
```

Comment:
instead of asking for cached diags, why not just ask for the diags appropriate for Document.  ANd if they have been cached they are returned, otherwise they are computed.

I'm very wary about this producing stale data that is then not updated properly as changed come in.

Reply:
Isn't this fairly high traffic, i.e.:

> if we simply changed this to "get the virtual project" again

Would mean we'd be continually launching processes in the background to re-ask for the content?